libs/mixer/melody.cpp

```diff
@@ -56,23 +60,85 @@ static int sineTone(uintptr_t userData, uint32_t position) {
     return position >= 512 ? -r : r;
 }
 
-static int sawtoothTone(uintptr_t userData, uint32_t position) {
-    (void)userData;
+static int sawtoothTone(uint8_t wave, uint32_t position, int channel, uint8_t cycle) {
+    (void)wave;
+    (void)channel;
+    (void)cycle;
     return (position << 6) - 0x7fff;
 }
 
-static int triangleTone(uintptr_t userData, uint32_t position) {
-    (void)userData;
+static int triangleTone(uint8_t wave, uint32_t position, int channel, uint8_t cycle) {
+    (void)wave;
+    (void)channel;
+    (void)cycle;
     return position < 512 ? (position << 7) - 0x7fff : ((1023 - position) << 7) - 0x7fff;
 }
 
-static int squareWaveTone(uintptr_t wave, uint32_t position) {
+static int squareWaveTone(uint8_t wave, uint32_t position, int channel, uint8_t cycle) {
+    (void)channel;
+    (void)cycle;
     return position < (102 * (wave - SW_SQUARE_10 + 1)) ? -0x7fff : 0x7fff;
 }
 
-static int silenceTone(uintptr_t userData, uint32_t position) {
-    (void)userData;
+static int tunedNoiseTone(uint8_t wave, uint32_t position, int channel, uint8_t cycle) {
+    // Generate a square wave filtered by a random bit sequence. Since the generator
+    // is called multiple times per wave, use static data to ensure we only generate
+    // a random bit once per wave, and then reuse it for future calls for that wave.
+    // Do this per channel to ensure results remain consistent when mixing multiple
```

Comment:
FYI, if you'd prefer to avoid passing the extra "channel" argument to all the generators, it should be possible to skip the per-channel tracking and use simple static state instead. This would mean that mixing multiple channels each playing tuned noise would have extra changes when switching channels. In the current setup, this would be a flipped bit every numsamples, so something like a 44Hz added noise assuming 44 KHz sample rate and 1024 sample buffers. This may well be inaudible, but I haven't been able to test that.

Reply:
Maybe we can pass PlayingSound to the generator function, and also add a word or two there for wave-form-data? Then maybe we can get rid of all that static state in these functions.